File: three_number_sum/three_number_sum.java
import java.util.*;

public class ThreeNumberSum {
    public static List<List<Integer>> threeNumberSum(int[] array, int target) {
        Arrays.sort(array);
        List<List<Integer>> tripleSums = new ArrayList<>();

        for (int i = 0; i < array.length - 2; i++) {
            int left = i + 1;
            int right = array.length - 1;

            while (left < right) {
                int currentSum = array[i] + array[left] + array[right];

                if (currentSum == target) {
                    tripleSums.add(Arrays.asList(array[i], array[left], array[right]));
                    left++;
                    right--;
                } else if (currentSum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return tripleSums;
    }

    public static void main(String[] args) {
        int[] input = {1, 2, 3, 4, 5, 6, -9, -10, 12};
        int target = 12;
        List<List<Integer>> result = threeNumberSum(input, target);
        System.out.println(result);
    }
}

File: three_number_sum/three_number_sum.go
package main

import (
	"fmt"
	"sort"
)

func threeNumberSum(array []int, target int) [][]int {
	sort.Ints(array)
	var tripleSums [][]int

	for i := 0; i < len(array)-2; i++ {
		left := i + 1
		right := len(array) - 1

		for left < right {
			currentSum := array[i] + array[left] + array[right]
			if currentSum == target {
				triplet := []int{array[i], array[left], array[right]}
				tripleSums = append(tripleSums, triplet)
				left++
				right--
			} else if currentSum < target {
				left++
			} else {
				right--
			}
		}
	}

	return tripleSums
}

func main() {
	input := []int{1, 2, 3, 4, 5, 6, -9, -10, 12}
	target := 12
	result := threeNumberSum(input, target)
	fmt.Println(result)
}
